File: backend/src/models/Materia.model.js
const mongoose = require("mongoose")

const {
  validaMediaParaPassar,
  validaPeso,
  validaNotaMaxima
} = require("../utils/validators.utils")

const {
  VALIDACAO
} = require("../constants/responseMessages.constants")

const MEDIA_DEFAULT = 6
const NOTA_MAXIMA_DEFAULT = 10

const materiaSchema = mongoose.Schema({
  nome: {
    type: String,
    required: [true, VALIDACAO.MATERIA.NOME_OBRIGATORIO],
    unique: true,
    trim: true
  },
  pesoProva: {
    type: Number,
    required: [true, VALIDACAO.MATERIA.PESO_PROVA_OBRIGATORIO],
    validate: {
      validator: validaPeso,
      message: VALIDACAO.MATERIA.PESO_DE_PROVA_INVALIDO
    }
  },
  pesoTrabalho: {
    type: Number,
    required: [true, VALIDACAO.MATERIA.PESO_TRABALHO_OBRIGATORIO],
    validate: {
      validator: validaPeso,
      message: VALIDACAO.MATERIA.PESO_DE_TRABALHO_INVALIDO
    }
  },
  mediaParaPassar: {
    type: Number,
    default: MEDIA_DEFAULT,
    validate: {
      validator: validaMediaParaPassar,
      message: VALIDACAO.MATERIA.MEDIA_PARA_PASSAR_INVALIDA
    }
  },
  notaMaxima: {
    type: Number,
    default: NOTA_MAXIMA_DEFAULT,
    validate: {
      validator: validaNotaMaxima,
      message: VALIDACAO.MATERIA.NOTA_MAXIMA_INVALIDA
    }
  }
})

const Materia = mongoose.model("Materia", materiaSchema)

module.exports = Materia

File: frontend/lib/scr/models/materia_model.dart
class Materia {
  final String? id;
  final String nome;
  final double pesoProva;
  final double pesoTrabalho;
  final double mediaParaPassar;
  final double notaMaxima;

  Materia({
    this.id,
    required this.nome,
    required this.pesoProva,
    required this.pesoTrabalho,
    required this.mediaParaPassar,
    required this.notaMaxima,
  });

  Map<String, dynamic> toJson() {
    return {
      'nome': nome,
      'pesoProva': pesoProva,
      'pesoTrabalho': pesoTrabalho,
      'mediaParaPassar': mediaParaPassar,
      'notaMaxima': notaMaxima,
    };
  }

  factory Materia.fromJson(Map<String, dynamic> json) {
    return Materia(
      id: json['_id'],
      nome: json['nome'],
      pesoProva: (json['pesoProva'] as num).toDouble(),
      pesoTrabalho: (json['pesoTrabalho'] as num).toDouble(),
      mediaParaPassar: (json['mediaParaPassar'] as num).toDouble(),
      notaMaxima: (json['notaMaxima'] as num).toDouble(),
    );
  }
}
